src/lazy/encoder/write_as_ion.rs

```diff
@@ -48,65 +48,47 @@ where
 
 // ===== WriteAsIonValue implementations for common types =====
 
-impl WriteAsIonValue for Null {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_null(self.0)
-    }
-}
-
-impl WriteAsIonValue for bool {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_bool(*self)
-    }
-}
-
-impl WriteAsIonValue for i32 {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_i64(*self as i64)
-    }
-}
-
-impl WriteAsIonValue for i64 {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_i64(*self)
-    }
-}
-
-impl WriteAsIonValue for usize {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_int(&Int::from(*self))
-    }
-}
-
-impl WriteAsIonValue for f32 {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_f32(*self)
-    }
-}
-
-impl WriteAsIonValue for f64 {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_f64(*self)
-    }
-}
-
-impl WriteAsIonValue for Decimal {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_decimal(self)
-    }
-}
-
-impl WriteAsIonValue for Timestamp {
-    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
-        writer.write_timestamp(self)
-    }
+macro_rules! impl_write_as_ion_value {
+    // End of iteration
+    () => {};
+    // The caller defined an expression to write other than `self` (e.g. `*self`, `*self.0`, etc)
+    ($target_type:ty => $method:ident with $self:ident as $value:expr, $($rest:tt)*) => {
+        impl WriteAsIonValue for $target_type {
+            fn write_as_ion_value<V: ValueWriter>(&$self, writer: V) -> IonResult<()> {
```

Comment:
The `$self` parameter must always be `self` for the trait implementation to be valid, right? And it has to be a macro parameter to get around the macro hygiene rules?

Reply:
Correct.